### PR TITLE
Introduce major version packet flags

### PR DESCRIPTION
**Pull request by mmedenjak**
Adds a packet flag on all outgoing packets. This is to support running a
cluster where a member can be connected to other members of the same
major version as well as have a connection to members of different major
versions. To distinguish between the two on a Packet level, we need to
introduce a Packet flag. This flag should be introduced on all versions
which will support migrating between 3.x and 4.x, which at this point
means 3.12.8+, 4.0.2+ and 4.1.x. The flags can be removed in 4.2 which
will not support data migration anymore.

Backport: https://github.com/hazelcast/hazelcast/pull/16992

### CODE REVIEW

**Comment by blazember** [comment]
Ok, `SendMemberHanshakeTask` exists only on master (4.1.z+) and is essentially the same as `BindRequest` from `4.0.z` that is covered in this PR. Ignore my last comment.

**Comment by vbekiaris** [comment]
minor question: is there a reason to introduce the `raiseFlags(FLAG_4_0)` method call around the code base instead of raising the flag in the constructors so all new `Packet`s are marked as from `4.0`? The `PacketDecoder` will `resetFlags` for incoming packets, so those coming with no flag set or with the 3_x flag will be handled appropriately.

Other than that the PR looks good

File: hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java [hunk 1]
```diff
@@ -43,6 +43,8 @@
     // 1. URGENT (bit 4)
     // 2. Packet type (bits 0, 2, 5)
     // 3. Flags specific to a given packet type (bits 1, 6)
+    // 4. 4.x flag (bit 7)
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java:46]: All packets in a stream will be of the same version. Why is this not part of the MemberHandshake?

File: hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java [hunk 2]
```diff
@@ -96,6 +98,16 @@
      */
     public static final int FLAG_JET_FLOW_CONTROL = 1 << 1;
 
+    /**
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java:101]: Why do we need 2 bits? One bit would be sufficient.

File: hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java:46]: It is part of the `BindRequest` (`MemberHandshake`). If you're asking why we don't just set the version on the tcp connection and use that, I guess the problem is getting a reference to the connection in all places where the Packet is passed around, and it introduces coupling between the connection/network layer and other subsystems like operation handling.

File: hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java (hunk 2 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java:101]: We have these special members both on 3.x and 4.x which can process _some_ packets both from 3.x and 4.x members. To do that, they rely on these flags. Let's say this special 4.x member says "if the FLAG_4_0 is absent, it's sent by 3.x and if it's present, it's sent by 4.x". This means, all members in the cluster that it's part of need to set this flag for the cluster to operate normally and this allows 3.8+ clusters to connect to this special member.

Now, we have the other side of the replication, for users wanting to replicate back from 4.x to 3.x. They need to run a special 3.12 member. To distinguish incoming packets from 3.12.x members and 4.x, we need to again check some flag. Let's say we don't have the FLAG_3_x and let's say, as you proposed we check "if the FLAG_4_0 is present, it's sent by 4.x and if it's absent, it's sent by 3.x". This places a restriction on the 4.x cluster -> it has to now be comprised out of members setting this flag, namely 4.0.2+.

So, to sum up, it's a decision on what kind of restriction we want to impose on users:
a) if we use just one flag like FLAG_4_0, the restriction would be "your 4.x cluster, be it the source or target of migration, has to be at least 4.0.2+"
b) if we use use two flags, the restriction would be "your target cluster which contains the special migration member has to be at least 4.0.2+ or 3.12.8+, depending on whether you're replicating 3.x->4.x or 4.x->3.x)

Also, I don't see using two flags as a big loss. Sure, it introduces a cognitive load but that's it. The bits are already there, they are just not used. Also, since migration tool will probably not support migrating to 4.2, we'll remove all traces of these flags in 4.2.

**Comment by mmedenjak** [comment]
@vbekiaris you're probably right about moving to the constructor, it just didn't occur to me. I'll do that then. Can't find where the `PacketDecoder` calls `resetFlags`, curious to see that, so I'll ping you about that.

File: hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java (hunk 2 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/nio/Packet.java:101]: In even shorter terms, the option with one flag places a restriction on the versions of both clusters while the option with two flags places a restriction on the versions of only one cluster.

**Comment by mmedenjak** [comment]
@vbekiaris addressed review comments

**Comment by mmedenjak** [comment]
Thank you for the reviews! Will send a forward-port to master now.
